peps/pep-0803.rst

```diff
@@ -0,0 +1,322 @@
+   The new limited API will not support CPython 3.14 and below.
+   Projects that need this support can build separate extensions specifically for
+   the 3.14 free-threaded interpreter, and for older stable ABI versions.
+
+API changes are OK
+   The new Limited API may require extension authors to make significant
+   changes to their code.
+   Projects that cannot do this (yet) can continue using Limited API 3.14,
+   which will yield extensions compatible with GIL-enabled builds only.
+
+No extra configuration
+   We do not introduce new “knobs” that influence what API is available
+   and what the ABI is compatible with.
+
+
+Specification
+=============
+
+
+Opaque PyObject
+---------------
+
+Version 3.15 of the Limited API will:
+
+- make the following structures *opaque* (or in C terminology, *incomplete
+  types*):
+
+  - :c:type:`PyObject`
+  - :c:type:`PyVarObject`
+  - :c:type:`!PyModuleDef_Base`
+  - :c:type:`PyModuleDef`
+
+- no longer include the following macros:
+
+  - :c:macro:`PyObject_HEAD`
+  - :c:macro:`!_PyObject_EXTRA_INIT`
+  - :c:macro:`PyObject_HEAD_INIT`
+  - :c:macro:`PyObject_VAR_HEAD`
+
+- no longer include these function-like macros:
+
+  - :c:func:`Py_SIZE`
+  - :c:func:`Py_SET_SIZE`
+  - :c:func:`Py_SET_TYPE`
+
+
+Implications
+^^^^^^^^^^^^
+
+Making the ``PyObject``, ``PyVarObject`` and ``PyModuleDef`` structures
+opaque means:
+
+- Their fields may not be accessed.
+
+  For example, instead of ``o->ob_type``, extensions must use
+  ``Py_TYPE(o)``.
+  This usage has been the preferred practice for some time.
+
+- Their size and alignment will not be available.
+  Expressions such as ``sizeof(PyObject)`` will no longer work.
+
+- They cannot be embedded in other structures.
+  This mainly affects instance structs of extension-defined types,
+  which will need to be defined using API added in :pep:`697` -- that is,
+  using a ``struct`` *without* ``PyObject`` (or other base class struct) at
+  the beginning, with :c:func:`PyObject_GetTypeData` calls needed to access
+  the memory.
+
+- Variables of these types cannot be created.
+  This mainly affects static ``PyModuleDef`` variables needed to define
+  extension modules.
+  Extensions will need to switch to API added in :pep:`793`.
+
+The following functions will become unusable in practice (in the new Limited
+API), since an extension cannot create valid, statically allocated, input
+for them. To ease the transition for extension developers, they will not yet be removed from the Limited API:
+
+- :c:func:`PyModuleDef_Init`
+- :c:func:`PyModule_Create`, :c:func:`PyModule_Create2`
+- :c:func:`PyModule_FromDefAndSpec`, :c:func:`PyModule_FromDefAndSpec2`
+
+
+New Export Hook (PEP 793)
+-------------------------
+
+Implementation of this PEP requires :pep:`793` (``PyModExport``:
+A new entry  point for C extension modules) to be
+accepted, providing a new “export hook” for defining extension modules.
+Using the new hook will become mandatory in Limited API 3.15.
```

Comment:
PEP 793 hasn't been accepted as at the time of writing. Would accepting this PEP require the SC to also accept 793?

Reply:
Yes. It's listed in the Requires header.